redis/src/main/scala/zio/redis/Interpreter.scala

```diff
@@ -1,135 +1,236 @@
-      readinessFlag: AtomicBoolean,
-      response: ByteBuffer
-    ) {
-      val receive: UIO[Any] =
-        UIO.effectSuspendTotal(if (readinessFlag.compareAndSet(true, false)) executeNext else UIO.unit)
-
-      def send(command: Chunk[String]): UIO[Promise[RedisError, String]] =
-        Promise.make[RedisError, String].flatMap(p => pendingRequests.offer(Request(command, p)).as(p))
-
-      private val executeNext: UIO[Any] =
-        pendingRequests.take.map { request =>
-          val exchange =
-            IO.effect {
-              try {
-                unsafeSend(request.command)
-                unsafeReceive()
-              } catch {
-                case t: Throwable => throw RedisError.ProtocolError(t.getMessage)
-              } finally readinessFlag.set(true)
-            }
+  }
 
-          request.result.unsafeDone(exchange.refineToOrDie[RedisError])
-        }
+  type ByteStream = Has[ByteStream.Service]
 
-      private def unsafeReceive(): String = {
-        val builder      = ArrayBuilder.make[Array[Byte]]
-        var readBytes    = 0
-        var responseSize = 0
+  object ByteStream {
 
-        // TODO: handle -1
-        while (readBytes == 0 || readBytes == ResponseBufferSize) {
-          channel.read(response)
-          response.flip()
+    val ResponseBufferSize = 1024
 
-          readBytes = response.remaining()
-          responseSize += readBytes
+    def connect: ZManaged[ByteStream, IOException, ReadWriteBytes] = ZManaged.accessManaged(_.get.connect)
 
-          val chunk = Array.ofDim[Byte](readBytes)
+    trait Service {
 
-          response.get(chunk)
+      def connect: Managed[IOException, ReadWriteBytes]
 
-          builder += chunk
+    }
 
-          response.clear()
-        }
+    trait ReadWriteBytes {
 
-        val chunks = builder.result()
-        val data   = Array.ofDim[Byte](responseSize)
-        var i      = 0
-        var j      = 0
+      def read: Stream[IOException, Byte]
 
-        while (i < chunks.length) {
-          val chunk = chunks(i)
+      def write(chunk: Chunk[Byte]): IO[IOException, Unit]
 
-          System.arraycopy(chunk, 0, data, j, chunk.length)
+    }
 
-          i += 1
-          j += chunk.length
+    /**
+     * Adapts `ComplectionHandler` async channel operations with interruption support.
+     */
+    private def effectAsyncChannel[C <: Channel, T](
+      channel: C
+    )(op: C => CompletionHandler[T, Any] => Any): IO[IOException, T] =
+      IO.effectAsyncInterrupt { k =>
+        val handler = new CompletionHandler[T, Any] {
+          def completed(result: T, u: Any): Unit = k(IO.succeedNow(result))
+
+          def failed(t: Throwable, u: Any): Unit =
+            t match {
+              case e: IOException => k(IO.fail(e))
+              case _              => k(IO.die(t))
+            }
         }
 
-        new String(data, UTF_8)
+        op(channel)(handler)
+        Left(IO.effect(channel.close()).ignore)
```

Comment:
Isn't the channel created in managed?

Reply:
This is for interruption. The only way to interrupt an in-progress I/O operation on a channel is to close the channel (well, the `java.util.concurrent.Future` version _technically_ lets you keep the channel open in a weird state, but it's cleaner just to close it).